### PR TITLE
Adding a Config Element to enable JNDI trace

### LINKED ISSUE #9902: Add ability to enable JNDI trace to ldapRegistry-3.0
It would be useful if we could enable a dump of the incoming and outgoing LDAP ASN.1 BER packets when using ldapRegsitry-3.0 to aid in resolving customer issues.

Add a new configuration element that would allow us to configure the `com.sun.jndi.ldap.trace.ber` environment property passed into the JNDI context.

The config would probably be something similar to the following, where the user points to a file. We probably also should support the use of "System.out".

`<ldapRegistry ... jndiBerOutputFile="${server.config.dir}/logs/ber-dump.log" ... />`

The code would look something like this:

```
if (berDumpEnabled) {
   env.put("com.sun.jndi.ldap.trace.ber", <FileOutputStream | System.out>);
}
```

### PR DESCRIPTION
**Pull request by agiridharan**
fixes #9902

### CODE REVIEW

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        updateConfigDynamically(libertyServer, server);
+
+    }
+
+    /**
+     *
+     * Check that we hit the timeout block of code for the context pool
+     */
+    @Test
+    public void testContextPoolOnTimeout() throws Exception {
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:211]: Update the name of the test, the test description and streamline it a bit -- does it really need a sleep?

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.impl.LibertyServerFactory;
+import componenttest.topology.utils.LDAPUtils;
+
+/**
+ * Ensure we're timing out the LDAP context pool when expected.
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:45]: Fix description

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:2]: Copyright

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/TimedDirContext.java [hunk 4]
```diff
@@ -49,6 +50,8 @@
     /** Constant for JNDI_CALL trace */
     public static final String JNDI_CALL = "JNDI_CALL ";
 
+    public static ByteArrayPlusOutputStream baos;
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/TimedDirContext.java:53]: Are we using this in this class?

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 5]
```diff
@@ -1056,6 +1068,16 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Set the LDAP read time out.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            System.out.println("Setting print stream");
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1075]: Remove or convert to trace.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/ByteArrayPlusOutputStream.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ * Copyright (c) 1997, 2004 IBM Corporation and others.
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/ByteArrayPlusOutputStream.java:2]: Fix copyright date

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/ByteArrayPlusOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    @Override
+    public void write(int b) {
+        super.write(b);
+        try {
+            if (this.toString("UTF-8").contains(eol)) {
+                try {
+                    super.writeTo(System.out);
+                    super.reset();
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/ByteArrayPlusOutputStream.java:37]: Fix all the e.printStackTrace -- we should trace it, but don't throw an exception. We don't want failing trace to interrupt an ldap reqeust.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 8]
```diff
@@ -1232,6 +1254,16 @@ public TimedDirContext reCreateDirContext(TimedDirContext oldCtx, String errorMe
...
     public void releaseDirContext(TimedDirContext ctx) throws WIMSystemException {
         final String METHODNAME = "releaseDirContext";
+        if (iJndiOutputEnabled) {
+            try {
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, METHODNAME + " Writing the rest of the packet capture");
+                }
+                bapos.writeTo(System.out);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1264]: Needs to be trace.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/publish/servers/com.ibm.ws.security.wim.adapter.ldap.fat.jndi.output/server.xml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<!--
+    Copyright (c) 2017 IBM Corporation and others.
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/publish/servers/com.ibm.ws.security.wim.adapter.ldap.fat.jndi.output/server.xml:2]: Copyright date

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/publish/servers/com.ibm.ws.security.wim.adapter.ldap.fat.jndi.output/bootstrap.properties [hunk 10]
```diff
@@ -0,0 +1,13 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/publish/servers/com.ibm.ws.security.wim.adapter.ldap.fat.jndi.output/bootstrap.properties:2]: Copyright date

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    /**
+     *
+     * Check that we hit the timeout block of code for the context pool
+     */
+    @Test
+    public void testContextPoolOnTimeout() throws Exception {
+        String trace = "-> localhost"; // depends on trace logged in LdapConnection.getDirContext
+        String returnUser = servlet.checkPassword(USER_DN, PWD);
+        List<String> errMsgs = libertyServer.findStringsInLogsAndTrace(trace);
+        assertTrue("Should find: " + trace, errMsgs.isEmpty());
```
Comment (kristip17) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:215]: "Should not find" ? You are making sure we don't log the trace, then enable it and we do find the trace?

**Comment by agiridharan** [comment]
#build

File: dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml [hunk 12]
```diff
@@ -639,6 +639,7 @@
     <dependency>
       <groupId>io.smallrye</groupId>
       <artifactId>smallrye-graphql</artifactId>
+
```
Comment (jvanhill) [inline dev/cnf/dependabot/check_this_in_if_it_changes/pom.xml:642]: This is weird.

File: dev/com.ibm.websphere.security.wim.base/src/com/ibm/websphere/security/wim/ConfigConstants.java [hunk 13]
```diff
@@ -402,6 +402,10 @@
      */
     String CONFIG_PROP_READ_TIMEOUT = "readTimeout";
 
+    /**
+     * Define whether JNDI BER packets will be written to the System.out
+     */
+    String CONFIG_PROP_JNDI_OUTPUT_ENABLED = "jndiOutputEnabled";
```
Comment (jvanhill) [inline dev/com.ibm.websphere.security.wim.base/src/com/ibm/websphere/security/wim/ConfigConstants.java:408]: add white space after.

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/metatype/metatype.xml [hunk 14]
```diff
@@ -57,6 +57,8 @@
             type="String" default="${count(sslRef)}"/>
         <AD id="SSLSupport.target" name="internal" description="internal use only"
             type="String" default="(repertoirePIDs=${sslRef})"/>
+		<AD id="jndiOutputEnabled" name="internal" description="internal use only"
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/metatype/metatype.xml:60]: Alignment.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Trivial;
+
+/******************************************************************************
+*	Use an extended version of ByteArrayOutputStream in order to allow access
+*	to protected values.
+******************************************************************************/
+@Trivial
+//rename BER output stream
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:28]: remove

```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+public class BEROutputStream extends ByteArrayOutputStream {
+    private static final TraceComponent tc = Tr.register(BEROutputStream.class);
+    public final String eol = getSystemProperty("line.separator");
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:31]: Use the method System.lineSeparator() instead.

```suggestion

```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public void write(int b) {
+        String METHODNAME = "write(int b)";
+        super.write(b);
+        try {
+            if (this.toString("UTF-8").contains(eol)) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:37]:
```suggestion
            if (this.toString("UTF-8").contains(System.lineSeparator())) {
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+                Tr.debug(tc, METHODNAME + " UTF-8 is not a valid encoding: " + e.toString());
+        }
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) {
+        String METHODNAME = "write(byte[] b, int off, int len)";
+        super.write(b, off, len);
+        try {
+            if (this.toString("UTF-8").contains(eol)) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:57]:
```suggestion
            if (this.toString("UTF-8").contains(System.lineSeparator())) {
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Convenience method to get a system property using
+     * {@link AccessController#doPrivileged(PrivilegedAction)}.
+     *
+     * @param property The property to retrieve.
+     * @return The value returned from {@link System#getProperty(String)}.
+     */
+    @Trivial
+    public static String getSystemProperty(final String property) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:85]: Delete this method. Use System.lineSeparator() instead.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java (hunk 18 above)
Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:57]: I also wonder if the BER output uses different line separators on windows vs unix? Have you tested on both?

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+import java.io.UnsupportedEncodingException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Trivial;
+
+/******************************************************************************
+*	Use an extended version of ByteArrayOutputStream in order to allow access
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/BEROutputStream.java:24]: Better Javadoc. Describe what the purpose of the class is.

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/metatype/metatype.xml [hunk 21]
```diff
@@ -57,6 +57,8 @@
...
+            required="false" type="Boolean"/>
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/OSGI-INF/metatype/metatype.xml:61]: default=false?

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 22]
```diff
@@ -1056,6 +1066,16 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Set the LDAP read time out.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1070]:
```suggestion
         * Enabled JNDI BER output if required.
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 23]
```diff
@@ -227,6 +235,8 @@
     /** Write to secondary server configuration. */
     private boolean iWriteToSecondary = false;
 
+    /** Buffer stream for catching JNDI output */
+    BEROutputStream bapos = new BEROutputStream();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:239]: Doesn't seem we should need this to be class level. Just add create it when you add it to the env.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 24]
```diff
@@ -1056,6 +1066,16 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Set the LDAP read time out.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            synchronized (iLock) {
+                PrintStream Stdout = new PrintStream(bapos);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1074]: Why do we need this in a PrintStream?

Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1074]: bapos should not be class level, create it here.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 25]
```diff
@@ -1232,6 +1252,18 @@ public TimedDirContext reCreateDirContext(TimedDirContext oldCtx, String errorMe
     @FFDCIgnore(NamingException.class)
     public void releaseDirContext(TimedDirContext ctx) throws WIMSystemException {
         final String METHODNAME = "releaseDirContext";
+        if (iJndiOutputEnabled) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1255]: We shouldn't need to do this here.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+        LdapRegistry ldap = new LdapRegistry();
+        ldap.setId("ldap1");
+        ldap.setRealm("LDAPRealm");
+        ldap.setHost("localhost");
+        ldap.setPort(String.valueOf(ds.getLdapPort()));
+        ldap.setBaseDN(SUB_DN);
+        ldap.setBindDN(InMemoryLDAPServer.getBindDN());
+        ldap.setBindPassword(InMemoryLDAPServer.getBindPassword());
+        ldap.setLdapType("Custom");
+        ContextPool cp = new ContextPool(true, 0, 2, 1, "2s", "5s");
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:195]: probably don't need to setup a context pool.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+        ldap.setContextPool(cp);
+        SearchResultsCache src = new SearchResultsCache();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:197]: probably should disable attributes cache as well. You want to hit your BEROutputStream as much as you can in the test.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+        ldap.setLdapCache(new LdapCache(null, src));
+
+        server.getLdapRegistries().add(ldap);
+        updateConfigDynamically(libertyServer, server);
+
+    }
+
+    /**
+     *
+     * Check that we hit the timeout block of code for the context pool
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:208]: This javadoc comment is outdated.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    @Test
+    public void testJNDIBEROutput() throws Exception {
+        String trace = "-> localhost"; // depends on trace logged in LdapConnection.getDirContext
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:212]: maybe search for a few items from the BER output if possible.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+        String returnUser = servlet.checkPassword(USER_DN, PWD);
+        List<String> errMsgs = libertyServer.findStringsInLogsAndTrace(trace);
+        assertTrue("Should not find: " + trace, errMsgs.isEmpty());
+
+        // Send in setting less than 1 second. Should run up to 1 second.
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:217]:
```suggestion
        /*
         * Enable JNDI BER trace output.
         */
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 31]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    /**
+     *
+     * Check that we hit the timeout block of code for the context pool
+     */
+    @Test
+    public void testJNDIBEROutput() throws Exception {
+        String trace = "-> localhost"; // depends on trace logged in LdapConnection.getDirContext
+        String returnUser = servlet.checkPassword(USER_DN, PWD);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:213]: /*
 * Run with JNDI BER trace disabled.
 */

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 32]
```diff
@@ -10,6 +10,7 @@
  *******************************************************************************/
 package com.ibm.ws.security.wim.adapter.ldap.context;
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:11]: Copyright date should be updated.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 33]
```diff
@@ -176,6 +181,8 @@
     /** The initial pool size for the DirContext pool. */
     private int iInitPoolSize = DEFAULT_INIT_POOL_SIZE;
 
+    private Boolean iJndiOutputEnabled = false;
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:184]: Javadoc

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 34]
```diff
@@ -1056,6 +1063,17 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Enabled JNDI BER output if required.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            synchronized (iLock) {
+                BEROutputStream berOS = new BEROutputStream();
+                PrintStream Stdout = new PrintStream(berOS);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1072]: Why do we need to wrap with a PrintStream? BEROutputStream is already an java.io.OutputStream.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConnection.java [hunk 35]
```diff
@@ -409,6 +410,8 @@ private void initializeContextManager(Map<String, Object> configProps) throws WI
          */
         iContextManager.setReadTimeout((Long) configProps.get(CONFIG_PROP_READ_TIMEOUT));
 
+        iContextManager.setJndiOutputEnabled((Boolean) configProps.get(CONFIG_PROP_JNDI_OUTPUT_ENABLED));
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/LdapConnection.java:413]: Comment?

File: dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/wim/LdapRegistry.java [hunk 36]
```diff
@@ -45,6 +45,7 @@
     private LdapFilters idsFilters;
     private Boolean ignoreCase;
     private LdapFilters iplanetFilters;
+    private Boolean jndiOutputEnabled;
```
Comment (jvanhill) [inline dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/wim/LdapRegistry.java:48]: Copyright date above should be 2020.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java [hunk 37]
```diff
@@ -0,0 +1,229 @@
...
+        ConfigElementList<LdapRegistry> ldaps = config.getLdapRegistries();
+        LdapRegistry ldap = null;
+        if (!ldaps.isEmpty() && ldaps.size() == 1) {
+            ldap = ldaps.get(0);
+        }
+        ldap.setJndiOutputEnabled(true);
+        updateConfigDynamically(libertyServer, config);
+        returnUser = servlet.checkPassword(USER_DN, PWD);
+        errMsgs = libertyServer.findStringsInLogsAndTrace(trace);
+        assertFalse("Should find: " + trace, errMsgs.isEmpty());
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:226]: I thought i put a comment about checking for more output, but I don't see it. Is there anything else we can check for? Maybe the incoming and  outgoing, maybe some data?

Comment (jvanhill) [reply dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/JNDIOutputTest.java:226]: Need to run with several JREs as well. the trace may be different.

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 38]
```diff
@@ -1056,6 +1063,18 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Enabled JNDI BER output if required.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            synchronized (iLock) {
+                BEROutputStream berOS = new BEROutputStream();
+                iEnvironment.put(LDAP_ENV_PROP_JNDI_BER_OUTPUT, berOS);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1072]:
```suggestion
                iEnvironment.put(LDAP_ENV_PROP_JNDI_BER_OUTPUT, berOS);
```
```suggestion
                iEnvironment.put(LDAP_ENV_PROP_JNDI_BER_OUTPUT, new BEROutputStream());
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 39]
```diff
@@ -1056,6 +1063,18 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Enabled JNDI BER output if required.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            synchronized (iLock) {
+                BEROutputStream berOS = new BEROutputStream();
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1071]:
```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 40]
```diff
@@ -1056,6 +1063,18 @@ public InitializeResult initialize() throws WIMApplicationException {
...
+                iEnvironment.put(LDAP_ENV_PROP_JNDI_BER_OUTPUT, berOS);
+//                PrintStream Stdout = new PrintStream(berOS);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1073]:
```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 41]
```diff
@@ -1056,6 +1063,18 @@ public InitializeResult initialize() throws WIMApplicationException {
...
+//                iEnvironment.put(LDAP_ENV_PROP_JNDI_BER_OUTPUT, Stdout);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1074]:
```suggestion
```

File: dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java [hunk 42]
```diff
@@ -1056,6 +1063,15 @@ public InitializeResult initialize() throws WIMApplicationException {
             iEnvironment.put(LDAP_ENV_PROP_READ_TIMEOUT, String.valueOf(DEFAULT_READ_TIMEOUT));
         }
 
+        /*
+         * Enabled JNDI BER output if required.
+         */
+        if (iJndiOutputEnabled != null && iJndiOutputEnabled) {
+            synchronized (iLock) {
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap/src/com/ibm/ws/security/wim/adapter/ldap/context/ContextManager.java:1070]: Did I miss this synchronized block? I don't think we need this. We are adding a new instance of the output stream to the environment, so I don't think there are any race conditions here.

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#build

**Comment by agiridharan** [comment]
#run-libby-bot

**Pull request merged**